### PR TITLE
Improvment/drag and drop

### PR DESCRIPTION
**Pull request by ydomenjoud**

### CODE REVIEW

File: projects/kalidea/kaligraphi/src/lib/atoms/kal-drag-drop/directives/kal-drop.directive.ts [hunk 1]
```diff
@@ -11,26 +13,63 @@ export interface KalDroppedEvent {
...
+
+  constructor(private draggingService: KalDragService) {
+  }
+
+  /**
+   * return Allowed position to drop on
+   */
+  @Input('kalDropPositions')
+  get positions() {
+    return Array.isArray(this._kalDropPositions) ? this._kalDropPositions : [];
```
Comment (jsp-up) [inline projects/kalidea/kaligraphi/src/lib/atoms/kal-drag-drop/directives/kal-drop.directive.ts]: coerceArray

File: projects/kalidea/kaligraphi/src/lib/atoms/kal-drag-drop/directives/kal-drop.directive.ts [hunk 2]
```diff
@@ -11,26 +13,63 @@ export interface KalDroppedEvent {
...
+  }
+
+  /**
+   * Allowed position to drop on
+   */
+  set positions(positions: KalDropPosition[]) {
+    this._kalDropPositions = positions;
```
Comment (jsp-up) [inline projects/kalidea/kaligraphi/src/lib/atoms/kal-drag-drop/directives/kal-drop.directive.ts]: coerceArray here too ?

File: projects/kalidea/kaligraphi/src/lib/atoms/kal-drag-drop/directives/kal-drop.directive.ts [hunk 3]
```diff
@@ -66,32 +113,89 @@ export class KalDropDirective implements OnDestroy {
...
-      this.dropPosition = KalDropPosition.Middle;
+    if (this.positions.length === 1) {
+      dropPosition = this.positions[0];
     } else {
-      this.dropPosition = null;
+      const positionFound = this
+        .buildPositionsConfig(targetHeight, positionsList)
+        .find(config => position < config.max && position > config.min);
+
+      if (positionFound) {
```
Comment (jsp-up) [inline projects/kalidea/kaligraphi/src/lib/atoms/kal-drag-drop/directives/kal-drop.directive.ts]: you can use a ternary

File: projects/kalidea/kaligraphi/src/lib/utils/decorators/auto-unsubscribe.ts [hunk 4]
```diff
@@ -29,7 +28,7 @@ export function AutoUnsubscribe() {
       // autoUnsubscribe
       target.constructor.prototype.ngOnDestroy = function () {
         target[decoratorSubscriptionsListKey].forEach((property) => {
-          const subscriptions: Subscription[] = isArray(this[property]) ? this[property] : [this[property]];
+          const subscriptions: Subscription[] = Array.isArray(this[property]) ? this[property] : [this[property]];
```
Comment (jsp-up) [inline projects/kalidea/kaligraphi/src/lib/utils/decorators/auto-unsubscribe.ts:31]: coerceArray

**Review by jsp-up — CHANGES_REQUESTED** [review]

File: projects/kalidea/kaligraphi/src/lib/atoms/kal-drag-drop/directives/kal-drop.directive.ts [hunk 5]
```diff
@@ -11,26 +13,63 @@ export interface KalDroppedEvent {
...
    * event emitted when user drop element
    */
   @Output() readonly kalDrop: EventEmitter<KalDroppedEvent> = new EventEmitter<KalDroppedEvent>();
 
   /**
-   * Allowed position to drop on
+   * callback to detect if element could be dropped on the current item
    */
-  @Input() kalDropPositions: KalDropPosition[] = [KalDropPosition.Top, KalDropPosition.Bot, KalDropPosition.Middle];
+  @Input() kalDropAllowed = undefined;
```
Comment (MichaelGrosio) [inline projects/kalidea/kaligraphi/src/lib/atoms/kal-drag-drop/directives/kal-drop.directive.ts]: @Input() kalDropAllowed

File: projects/kalidea/kaligraphi/src/lib/atoms/kal-drag-drop/services/kal-drag.service.ts [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+/**
+ * @TODO manage multitouch screen with multi drag
+ */
+
+@Injectable({
+  providedIn: 'root'
+})
+export class KalDragService {
+
+  private draggingSubject$ = new BehaviorSubject(null);
```
Comment (MichaelGrosio) [inline projects/kalidea/kaligraphi/src/lib/atoms/kal-drag-drop/services/kal-drag.service.ts:13]: type the value ?

**Review by MichaelGrosio — CHANGES_REQUESTED** [review]

File: projects/kalidea/kaligraphi/src/lib/atoms/kal-drag-drop/directives/kal-drop.directive.ts [hunk 7]
```diff
@@ -66,32 +113,89 @@ export class KalDropDirective implements OnDestroy {
...
+
+    // update config if we have less than 3 positions
+    if (positions.length === 1) {
+      // if we have only one drop position available, set it as default on hover
+      positions[0].min = 0;
+      positions[0].max = targetHeight;
+    } else if (positions.length === 2) {
+      // if we have two drop positions available, distribute remaining space
+      if (!this.isPositionAvailable(KalDropPosition.Top)) {
+        positions.find(config => config.position === KalDropPosition.Middle).min = 0;
```
Comment (MichaelGrosio) [inline projects/kalidea/kaligraphi/src/lib/atoms/kal-drag-drop/directives/kal-drop.directive.ts]: const positionMiddle = positions.find(config => config.position === KalDropPosition.Middle)

File: projects/kalidea/kaligraphi/src/lib/atoms/kal-drag-drop/directives/kal-drop.directive.ts [hunk 8]
```diff
@@ -66,32 +113,89 @@ export class KalDropDirective implements OnDestroy {
...
+  })
+  private buildPositionsConfig(targetHeight: number, positionsList: KalDropPosition[]): KalDropPositionConfig[] {
+
+    const positions: KalDropPositionConfig[] = [];
+
+    const top = targetHeight * this.kalDropThreshold;
+    const bot = targetHeight * (1 - this.kalDropThreshold);
+
+    // list of defaut config
+    const positionsConfig: { [key: string]: { min: number, max: number } } = {
```
Comment (MichaelGrosio) [inline projects/kalidea/kaligraphi/src/lib/atoms/kal-drag-drop/directives/kal-drop.directive.ts:144]: you can move this in method

**Review by MichaelGrosio — CHANGES_REQUESTED** [review]

File: projects/kalidea/kaligraphi/src/lib/atoms/kal-drag-drop/services/kal-drag.service.ts (hunk 6 above)
Comment (ydomenjoud) [reply projects/kalidea/kaligraphi/src/lib/atoms/kal-drag-drop/services/kal-drag.service.ts:13]: can't set type, it can be anything

**Review by MichaelGrosio — APPROVED** [review]

**Review by jsp-up — APPROVED** [review]

**Pull request merged**